File: implementations/go/examples/internal/application/event/classify_refund_reason_event_handler.go
package event

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"

	"github.com/example/account-service/internal/domain/payment"
)

// ClassifyRefundReasonEventHandler reacts to RefundRequested (published
// unconditionally by payment.NewRefund, before EvaluateRefundEligibility's
// approve/reject judgment even runs) to classify the refund's free-text
// reason for ops-analytics reporting only — see
// application/query/get_refund_reason_insights_handler.go. Runs off the
// request hot path (RequestRefundHandler never calls an LLM directly), and
// its result is never read back into any eligibility/approval decision.
// Inherently idempotent: a retried delivery just re-runs the same
// find→classify→save cycle, landing on the same (or an equally acceptable)
// category.
type ClassifyRefundReasonEventHandler struct {
	classifier RefundReasonClassifier
	refunds    payment.RefundRepository
}

func NewClassifyRefundReasonEventHandler(classifier RefundReasonClassifier, refunds payment.RefundRepository) *ClassifyRefundReasonEventHandler {
	return &ClassifyRefundReasonEventHandler{classifier: classifier, refunds: refunds}
}

// Handle satisfies the outbox.Handler signature — it is invoked whenever the
// Consumer encounters an event_type="RefundRequested" message.
func (h *ClassifyRefundReasonEventHandler) Handle(ctx context.Context, payload []byte) error {
	var evt payment.RefundRequested
	if err := json.Unmarshal(payload, &evt); err != nil {
		return fmt.Errorf("unmarshal RefundRequested: %w", err)
	}

	refund, err := payment.FindOneRefund(ctx, h.refunds, evt.RefundID)
	if err != nil {
		if errors.Is(err, payment.ErrNotFound) {
			// A stale/duplicate delivery referencing a since-deleted refund
			// degrades to a no-op instead of blocking retry forever — the
			// same posture as CategorizeTransactionEventHandler's
			// nil-transaction check.
			return nil
		}
		return fmt.Errorf("find refund: %w", err)
	}

	category := h.classifier.Classify(ctx, evt.Reason)
	refund.CategorizeReason(category)

	if err := h.refunds.SaveRefund(ctx, refund); err != nil {
		return fmt.Errorf("save refund: %w", err)
	}
	slog.InfoContext(ctx, "refund reason classified", "refund_id", evt.RefundID, "category", category)
	return nil
}

File: implementations/fastapi/examples/src/payment/application/event/classify_refund_reason_event_handler.py
from __future__ import annotations

import logging

from ...domain.refund_repository import RefundRepository
from ..service.refund_reason_classifier import RefundReasonClassifier

logger = logging.getLogger(__name__)


class ClassifyRefundReasonEventHandler:
    """Reacts to RefundRequested (published unconditionally by Refund.create(), before
    RefundEligibilityService's approve/reject judgment even runs) to classify the refund's
    free-text reason asynchronously, off the request hot path (RequestRefundHandler never
    calls an LLM directly) — purely for ops-analytics reporting, see
    refund_reason_insights_query.py. Its result is never read back into any eligibility/
    approval decision. Inherently idempotent (see domain-events.md): a retried delivery just
    re-runs the same find -> classify -> save cycle, landing on the same (or an equally
    acceptable) category.
    """

    def __init__(self, classifier: RefundReasonClassifier, refund_repo: RefundRepository) -> None:
        self._classifier = classifier
        self._refund_repo = refund_repo

    async def handle(self, payload: dict) -> None:
        refunds, _ = await self._refund_repo.find_refunds(page=0, take=1, refund_id=payload["refund_id"])
        refund = refunds[0] if refunds else None
        if refund is None:
            return

        category = await self._classifier.classify(payload["reason"])
        refund.categorize_reason(category)
        await self._refund_repo.save_refund(refund)
        logger.info("Refund reason classified: refund_id=%s category=%s", payload["refund_id"], category)
